### ISSUE TITLE
salt.runner + state.orchestrate + parallel: True - is not parallel

### ISSUE THREAD

**Issue by Oloremo** [saltstack/salt#55121]
### Description of Issue
I'm trying to write an orchestration that could be executed in parallel when possible. And even I set `parallel: True` it executed serially.

### Setup
I'm trying to do something like this: https://gist.github.com/Oloremo/f786f71044233e0db423d1d873e6c4ab

### Steps to Reproduce Issue
I prepared a script to create salt orch structure: https://gist.github.com/Oloremo/90b600f60c48da051023a33da50ce5f7

cd to your salt file_root. and run the provided script.
Directory `delme` will be created in it and populated with 6 files needed for parallel test.
After you're free to run something like
`salt-run state.orchestrate delme.service_orch -l debug 2>&1 | egrep 'Running state|Completed state'`

**Expected**: serviceB and serviceC to be run in parallel.
**Reality**: All services run serially one after another.

At first, I thought it's because I'm running 2 states on a single minion. But feel free to modify targets from the example above so **serviceB** and **serviceC** will be run on different minions - and the results will be the same.

### Versions Report
```
Salt Version:
           Salt: 2019.2.0

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.10.1
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.6.2
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.6.8 (default, Aug  7 2019, 17:28:10)
   python-gnupg: 0.4.5
         PyYAML: 5.1.2
          PyZMQ: 18.0.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.1

System Versions:
           dist: centos 7.5.1804 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-862.3.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.5.1804 Core
```

**Comment by smartass101** [comment]
Searching through the docs I came accross [salt.states.saltmod.parallel_runners](https://docs.saltstack.com/en/latest/ref/states/all/salt.states.saltmod.html#salt.states.saltmod.parallel_runners) which has a docstring that suggests that it is meant to be used in orchestration.
>This state is intended to be used with the orchestrate runner in place of the saltmod.runner state when different tasks should be run in parallel.

Perhaps this is an issue of lacking documentation in the Orchestration part of docs.

On the other hand, discussion in https://github.com/saltstack/salt/issues/50324#issuecomment-434753418 seems to suggest that `parallel_runners` should be deprecated in favor of `parallel: True` as it is deemed to work.

**Comment by Oloremo** [comment]
From my understanding - both `parallel_runners` and runner with parallel: True should work and do the same as for now. 

I tried to use a `parallel_runners` and it's broken for test mode right now and yeah seems like nobody really supports it so I stick with the runner + parallel: True

**Comment by Oloremo** [comment]
i'm pretty sure it's somehow related to requisites I think I saw some related issue...

**Comment by smartass101** [comment]
My understanding from the [parallel docs](https://docs.saltstack.com/en/latest/ref/states/parallel.html#parallel-states-and-requisites) is that the requisites should be defined first, because salt still evaluates the states in order of appearance in the file and branches off sub-processes in that order. I.e. in your example I think `java` should be defined first and `hbase` the last. And `java` `hbase` perhaps would not need `parallel: True` in that case. It's actually only `zookeeper` and `hadoop` that could potentially benefit from `parallel: True` in your example.

**Comment by Oloremo** [comment]
Order in file should not matter if I'm using requisites since requisites will re-order state graph based on them. And regarding benefit - I get it, but I'm trying to make a generalized approach.

**Comment by Oloremo** [comment]
Probably the root cause: https://github.com/saltstack/salt/issues/49273

**Comment by Oloremo** [comment]
Stale bot go away, this is an issue

**Comment by xeacott** [comment]
I'll also bump this pending related changes to https://github.com/saltstack/salt/issues/49273

**Comment by xeacott** [comment]
Apologies for the delay @Oloremo

**Comment by gabrielgt** [comment]
I tried the code made by @garethgreenaway in PR #58976 that is probably related but unfortunately it does not fix this issue. May be @garethgreenaway knows how to fix this issue too :)

**Comment by Oloremo** [comment]
I just tried to run my example script using the Salt master and I had to move serviceB and serviceC on a different nodes and it seems to work as expected!

```
$ salt-run state.orchestrate delme.service_orch -l debug 2>&1 | egrep 'Running state|Completed state'
[INFO    ] Running state [state.orchestrate] at time 11:38:26.874773
[INFO    ] Completed state [state.orchestrate] at time 11:38:26.881244 (duration_in_ms=6.474)
[INFO    ] Running state [orch.serviceA] at time 11:38:40.021707
[INFO    ] Completed state [orch.serviceA] at time 11:39:24.210257 (duration_in_ms=44188.55)
[INFO    ] Running state [state.orchestrate] at time 11:39:24.227174
[INFO    ] Completed state [state.orchestrate] at time 11:39:24.232396 (duration_in_ms=5.222)
[INFO    ] Running state [state.orchestrate] at time 11:39:24.233473
[INFO    ] Completed state [state.orchestrate] at time 11:39:24.238478 (duration_in_ms=5.005)
[INFO    ] Running state [orch.serviceB] at time 11:39:37.769900
[INFO    ] Running state [orch.serviceC] at time 11:39:38.635188
[INFO    ] Completed state [orch.serviceB] at time 11:40:14.952326 (duration_in_ms=37182.426)
[INFO    ] Completed state [orch.serviceC] at time 11:40:15.165351 (duration_in_ms=36530.164)
[INFO    ] Running state [state.orchestrate] at time 11:40:15.175794
[INFO    ] Completed state [state.orchestrate] at time 11:40:15.180594 (duration_in_ms=4.8)
[INFO    ] Running state [orch.serviceD] at time 11:40:28.419675
[INFO    ] Completed state [orch.serviceD] at time 11:41:11.192799 (duration_in_ms=42773.124)
```

**Comment by Oloremo** [comment]
So I'm closing this issue as resolved. Thanks!

**Comment by gabrielgt** [comment]
Hello @Oloremo, first of all thanks for your help. I tried your example script witch v.3003.rc1 and it worked, but mine did not. So, I looked for the difference. I have modified your script to reproduce the error with minimal changes. Basically you did:
```
A __ B __ D
  \_ C _/
```
And it worked, but with the following modification it failed:
```
A __ B1 __ B2 __ D
  \_ C _________/
```
In this example, B1 and C should start at the same time, but C starts with B2 instead. Because the duration of C is now 60 seconds, it lasts 150 seconds in total, instead of 120:
```
30 __ 30 __ 30 __________30
         \_ 60 _________/
```

Of course, the requirements between tasks can be much more complicated, but let's start with the simple.
The new script can be downloaded from https://gist.github.com/gabrielgt/e6c5337e2bd69dfefc61cee2c06cc12c

The result was the following:
```
$time sudo salt-run state.orch delme.service_orch -l debug 2>&1 | egrep 'Running state|Completed state'
[INFO    ] Running state [state.orchestrate] at time 18:32:14.612501
[INFO    ] Completed state [state.orchestrate] at time 18:32:14.620257 (duration_in_ms=7.748)
[INFO    ] Running state [orch.serviceA] at time 18:32:17.827567
[INFO    ] Completed state [orch.serviceA] at time 18:32:49.561936 (duration_in_ms=31734.369)
[INFO    ] Running state [state.orchestrate] at time 18:32:49.584701
[INFO    ] Completed state [state.orchestrate] at time 18:32:49.600972 (duration_in_ms=16.271)
[INFO    ] Running state [orch.serviceB1] at time 18:32:53.885257
[INFO    ] Completed state [orch.serviceB1] at time 18:33:25.768963 (duration_in_ms=31883.705)
[INFO    ] Running state [state.orchestrate] at time 18:33:25.789756
[INFO    ] Completed state [state.orchestrate] at time 18:33:25.804843 (duration_in_ms=15.088)
[INFO    ] Running state [state.orchestrate] at time 18:33:25.807449
[INFO    ] Completed state [state.orchestrate] at time 18:33:25.831102 (duration_in_ms=23.653)
[INFO    ] Running state [orch.serviceC] at time 18:33:30.419105
[INFO    ] Running state [orch.serviceB2] at time 18:33:31.251604
[INFO    ] Completed state [orch.serviceB2] at time 18:34:03.667342 (duration_in_ms=32415.737)
[INFO    ] Completed state [orch.serviceC] at time 18:34:32.920972 (duration_in_ms=62501.866)
[INFO    ] Running state [state.orchestrate] at time 18:34:32.943566
[INFO    ] Completed state [state.orchestrate] at time 18:34:32.956901 (duration_in_ms=13.336)
[INFO    ] Running state [orch.serviceD] at time 18:34:36.844001
[INFO    ] Completed state [orch.serviceD] at time 18:35:08.826227 (duration_in_ms=31982.226)

real    2m59.788s
```

**Comment by Oloremo** [comment]
@gabrielgt Sad :-(

Mind to raise another issue and link it to this one and to the other mentioned here?
